keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,464 @@
+very similar to what is done today since validation for updates is run AFTER all
+mutations.
+
+### Where does this happen
+
+This process is meant to happen right before the update validation and after
+mutating, but before validating webhooks, and only run on updates. This will allow us to
+keep the exact same behavior while removing the validation code that checks the
+immutability of fields.
+
+![Decoding steps which must apply defaults](20191001-crd-immutability-pipeline.png)
+
+### Risks and Mitigations
+
+- immutable metadata would break API machinery. We forbid the `x-kubernetes-immutable`
+  at the root of the object and inside `.metadata`. `kind` and `apiVersion` are
+  immutable implicitly. We might publish immutable though for some of these fields.
+
+## Design Details
+
+### Test Plan
+
+- exhaustive unit tests are added in apiextensions-apiserver for
+  - CRD validation
+    - for `x-kubernetes-immutable: true` and `x-kubernetes-immutable-keys: true` at the root and in `metadata`
+    - for `x-kubernetes-immutable-keys: true` for list-type `map` and forbidden otherwise
+    - for `x-kubernetes-immutable: true` on the list-keys of `x-kubernetes-immutable-keys: true` maps.
+    - for `x-kubernetes-immutable: true` and `x-kubernetes-immutable-keys: true` only for v1 CRDs, or during ratcheting updates.
+  - immutability checking with all variants of `x-kubernetes-immutable`, `x-kubernetes-immutable-keys` and `x-kubernetes-list-type`.
+- integration tests are added for
+  - creation, updates, patches and server-side-apply of partially immutable CRs 
+  - interaction of server-side-apply list-types and immutability
+  - OpenAPI publishing of the vendor extensions
+  - CRD updates of the immutability extensions and that the new immutability
+    schemas are followed.
+  - rejection of `x-kubernetes-immutable: true` and `x-kubernetes-immutable-keys: true` for non-v1 CRDs
+- e2e and conformance tests that
+  - immutability is followed during updates, patches and server-side-apply. 
+   
+### Graduation Criteria
+
+Because of the very limited risk of the additional immutability check and our
+experience with defaulting getting no additional soak time by the community
+we propose to start directly as beta.
+
+For beta:
+
+- performance does not suffer for CRDs **which do not use** immutability vendor extensions.
+
+For GA:
+
+- performance is benchmarked with an upper bound overhead of 15% on CRDs with schemas.
+
+## Implementation History
+
+N/A
+
+## Alternative Considered
+
+- OpenAPI has a notion of `readOnly`. This is meant to restrict fields to be set
+  only in responses, not in a request payload. This does not match our 
+  `never-change-after-creation` semantics.
+- Allowing `false` as value for `x-kubernetes-immutable: false` was considered to
+  disable immutability imposed by a parent node. This complicates the semantics
+  considerably and can be expressed with a combination of `x-kubernetes-immutable-keys` 
```

Comment:
This sentence contradicts itself. If you can already express it, the system can mechanically convert one expression to the other, the semantics must already be the same.